chat-android/src/main/java/com/ably/chat/PaginatedResult.kt

```diff
@@ -0,0 +1,65 @@
+package com.ably.chat
+
+import com.google.gson.JsonElement
+import io.ably.lib.types.AblyException
+import io.ably.lib.types.AsyncHttpPaginatedResponse
+import io.ably.lib.types.ErrorInfo
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+/**
+ * Represents the result of a paginated query.
+ */
+interface PaginatedResult<T> {
+
+    /**
+     * The items returned by the query.
+     */
+    val items: List<T>
+
+    /**
+     * Fetches the next page of items.
+     */
+    suspend fun next(): PaginatedResult<T>
+
+    /**
+     * Whether there are more items to query.
+     *
+     * @returns `true` if there are more items to query, `false` otherwise.
+     */
+    fun hasNext(): Boolean
+}
+
+fun <T> AsyncHttpPaginatedResponse?.toPaginatedResult(transform: (JsonElement) -> T): PaginatedResult<T> =
+    this?.let { AsyncPaginatedResultWrapper(it, transform) } ?: EmptyPaginatedResult()
```

Comment:
when would `EmptyPaginatedResult()` get returned?

Reply:
Right now when there is no body, SDK return `null`. I don't like this because we already return container, that can be empty, so I return `EmptyPaginatedResult` instead. For end user there is no difference between empty list and empty body, the same behavior we have in js. 

The other case is when we invoke `next()` and `hasNext()` is `false`. I personally would throw exception in this case, but decided not to do this, because we don't throw exception in core SDK, and silently return `null`. Once again don't want to return `null`, because you definitely don't want to do null-checks here.

Reply:
Interesting - I personally think returning null makes more sense when looking at the code but I am coming from an optional-centric (and crazy) language in Swift. This will likely be one of the points in which we deviate between Kotlin and Swift due to platform best practices

Reply:
Not sure it's Kotlin best practice, the best practice probably would be throwing exceptions, but I decide to silence them as Ably Java does. This will be an extremely small change. 

And about optional-centricity: drom the SDK user’s perspective,  I personally  don’t think `PaginatedResult#next()` should return a nullable result, both for Swift and Kotlin—I’ve never seen this before, except in Ably Java. :) Users will have to handle this nullability on their own, but there is no reason for that.